internal/measurement/measurement.go

```diff
@@ -262,67 +242,104 @@ func timeLabel(value int64, fromUnit, toUnit string) (v float64, u string, ok bo
+	fromUnit = strings.TrimSuffix(strings.ToLower(fromUnit), "s")
+	toUnit = strings.TrimSuffix(strings.ToLower(toUnit), "s")
+	uf := func(unit string) (string, float64, bool) {
+		return unitFactor(unit, gcuUnits)
+	}
+	as := func(value float64) (float64, string, bool) {
+		return autoScale(value, gcuUnits)
+	}
+	return convertUnit(value, fromUnit, toUnit, unit{"GCU", []string{}, 1.0}, uf, as)
+}
+
+// convertUnit converts a value from the fromUnit to the toUnit, autoscaling
+// the value if the toUnit is "minimum" or "auto". If the fromUnit is not
+// included in units, then a false boolean will be returned. If the toUnit
+// is not in units, the value will be returned in terms of the default unit.
+func convertUnit(value int64, fromUnit, toUnit string, defaultUnit unit, unitFactor func(string) (string, float64, bool), autoScale func(float64) (float64, string, bool)) (v float64, u string, ok bool) {
+	_, fromUnitFactor, ok := unitFactor(fromUnit)
+	if !ok {
 		return 0, "", false
 	}
 
+	v = float64(value) * fromUnitFactor
+
 	if toUnit == "minimum" || toUnit == "auto" {
-		switch {
-		case d < 1*time.Microsecond:
-			toUnit = "ns"
-		case d < 1*time.Millisecond:
-			toUnit = "us"
-		case d < 1*time.Second:
-			toUnit = "ms"
-		case d < 1*time.Minute:
-			toUnit = "sec"
-		case d < 1*time.Hour:
-			toUnit = "min"
-		case d < 24*time.Hour:
-			toUnit = "hour"
-		case d < 15*24*time.Hour:
-			toUnit = "day"
-		case d < 120*24*time.Hour:
-			toUnit = "week"
-		default:
-			toUnit = "year"
+		if v, u, ok := autoScale(v); ok {
+			return v, u, true
 		}
+		return v / defaultUnit.factor, defaultUnit.canonicalName, true
 	}
 
-	var output float64
-	dd := float64(d)
-	switch toUnit {
-	case "ns", "nanosecond":
-		output, toUnit = dd/float64(time.Nanosecond), "ns"
-	case "us", "microsecond":
-		output, toUnit = dd/float64(time.Microsecond), "us"
-	case "ms", "millisecond":
-		output, toUnit = dd/float64(time.Millisecond), "ms"
-	case "min", "minute":
-		output, toUnit = dd/float64(time.Minute), "mins"
-	case "hour", "hr":
-		output, toUnit = dd/float64(time.Hour), "hrs"
-	case "day":
-		output, toUnit = dd/float64(24*time.Hour), "days"
-	case "week", "wk":
-		output, toUnit = dd/float64(7*24*time.Hour), "wks"
-	case "year", "yr":
-		output, toUnit = dd/float64(365*24*time.Hour), "yrs"
-	default:
-		// "sec", "second", "s" handled by default case.
-		output, toUnit = dd/float64(time.Second), "s"
+	toUnit, toUnitFactor, ok := unitFactor(toUnit)
+	if !ok {
+		return v / defaultUnit.factor, defaultUnit.canonicalName, true
+	}
+	return v / toUnitFactor, toUnit, true
+}
+
+// unitFactor returns the canonical name for the unit, the factor by which one
+// must multiply a value specified in terms of unit in order to get the value
+// specified in terms of the base unit, and a boolean to indicated if a unit
+// was identified for the specified unit within the slice units.
+func unitFactor(unit string, units []unit) (string, float64, bool) {
```

Comment:
This is optional, but I would consider a couple of things API-wise, see WDYT.

We could make []unit a type so that it can be a receiver. I'd also name unitFactor as findByAlias and return *unit from it - the reason is that it returns more than just the factor, it pretty much searches for a unit by alias.

Having a knownUnit function could simplify isTimeUnit / isMemoryUnit / isGCUUnit a bit.

So, something like:

```
type unitList []unit
...
// findByAlias .... Returns nil if the unit with such alias is not found.
func (us unitList) findByAlias(alias string) *unit {
  ...
}

func (us unitList) autoScale(...) ... {
  ...
}

func (us unitList) sniffUnit(unit string) *unit {
  return us.findByAlias(strings.TrimSuffix(strings.ToLower(unit), "s"))
}
```

With this, I think you could remove isTimeUnit and its friends as it simply becomes `if timeUnits.sniffUnit(s) != nil`.

I think `convertUnit` would also simplify, since now it can also accept a receiver and it doesn't need to accept the function pointers. I could be wrong on this, but I feel that this can simplify.

Reply:
Thanks! This is much simpler. 

One thing that happened with this change: 'cycle' is no longer treated as a time unit. There weren't any tests which failed because of this, but wanted to call it out in case this could be a problem.